src/producer.py

```diff
@@ -0,0 +1,185 @@
+import asyncio
+import os
+import random
+import struct
+from datetime import datetime
+
+import typer
+from fastapi import FastAPI
+from prometheus_client import Counter, Histogram, make_asgi_app
+from streamstore import S2, Stream
+from streamstore.schemas import AppendInput, Record
+from uvicorn import Config, Server
+
+S2_AUTH_TOKEN = os.getenv("S2_AUTH_TOKEN")
+
+CYCLE_EVERY = 0.5  # 0.5s or 500ms
+MAX_JITTER = 0.2  # 0.2s or 200ms
+TIMESTAMP_HEADER_NUM_BYTES = 10
+
+METRICS_NAMESPACE = "stress"
+
+
+appended_records_counter = Counter(
+    name="appended_records_total",
+    documentation="Total number of records successfully appended.",
+    namespace=METRICS_NAMESPACE,
+)
+appended_bytes_counter = Counter(
+    name="appended_bytes_total",
+    documentation="Total number of bytes successfully appended.",
+    namespace=METRICS_NAMESPACE,
+)
+append_latency_histogram = Histogram(
+    name="append_latency_seconds",
+    documentation="Time taken for each append operation in seconds.",
+    namespace=METRICS_NAMESPACE,
+)
+append_errors_counter = Counter(
+    name="append_errors_total",
+    documentation="Total number of failed append operations.",
+    namespace=METRICS_NAMESPACE,
+)
+
+
+async def append(stream: Stream, input: AppendInput) -> None:
+    start = datetime.now()
+    with append_errors_counter.count_exceptions():
+        output = await stream.append(input)
+        appended_bytes_counter.inc(
+            sum(
+                len(record.body) + TIMESTAMP_HEADER_NUM_BYTES
+                for record in input.records
+            )
+        )
+        appended_records_counter.inc(output.end_seq_num - output.start_seq_num)
+    elapsed = datetime.now() - start
+    append_latency_histogram.observe(elapsed.total_seconds())
+
+
+async def produce_records(
+    stream: Stream,
+    batch_size: int,
+    record_body_size: int,
+    per_cycle_target_num_bytes: int,
+):
+    cur_cycle_num_bytes = 0
+    append_coros = []
+    while cur_cycle_num_bytes < per_cycle_target_num_bytes:
+        records = []
+        for _ in range(random.choice(batch_size)):
+            record = Record(
+                body=os.urandom(random.choice(record_body_size)),
+                headers=[(b"ts", struct.pack("d", datetime.now().timestamp()))],
```

Comment:
Maybe just stringified timestamp as utf-8 bytes?

Reply:
just remembered why i went this one. just to have a round 10 bytes header.